File: test-automation-core/src/main/java/com/nortal/test/core/plugin/JsonPerFeatureFormatterPlugin.java
package com.nortal.test.core.plugin;

import io.cucumber.plugin.EventListener;
import io.cucumber.plugin.event.EmbedEvent;
import io.cucumber.plugin.event.EventPublisher;
import io.cucumber.plugin.event.TestCaseFinished;
import io.cucumber.plugin.event.TestCaseStarted;
import io.cucumber.plugin.event.TestSourceRead;
import io.cucumber.plugin.event.TestStepFinished;
import io.cucumber.plugin.event.TestStepStarted;
import io.cucumber.plugin.event.WriteEvent;

/**
 *
 * This class is a cucumber plugin that outputs tests results in a json format.
 * The main difference between this and standard json plugin is that this one will create a separate json file per feature.
 * <p>
 * Usage: plugin should be used as any other cucumber plugin by passing it to the runner.
 * <pre><code>plugin = { "com.nortal.test.commons.plugin.JsonPerFeatureFormatterPlugin:test-output/cucumber-report" }</code></pre>
 * </p>
 * Output folder should be provided by passing in a relative path as a parameter (the part after colon)
 */
public class JsonPerFeatureFormatterPlugin implements EventListener {

    private JsonPerFeatureReporter reporter;

    public JsonPerFeatureFormatterPlugin(String outputPath) {
        this.reporter = new JsonPerFeatureReporter(outputPath);
    }

    @Override
    public void setEventPublisher(EventPublisher publisher) {
        publisher.registerHandlerFor(TestSourceRead.class, event -> reporter.readSources(event));
        publisher.registerHandlerFor(TestCaseStarted.class, event -> reporter.addTestCase(event));
        publisher.registerHandlerFor(TestStepStarted.class, event -> reporter.startStep(event));
        publisher.registerHandlerFor(TestStepFinished.class, event -> reporter.finishStep(event));
        publisher.registerHandlerFor(TestCaseFinished.class, event -> reporter.finishTestCase(event));
        publisher.registerHandlerFor(WriteEvent.class, event -> reporter.addText(event));
        publisher.registerHandlerFor(EmbedEvent.class, event -> reporter.embed(event));
    }

}

File: test-automation-core-old/src/main/java/com/nortal/test/core/plugin/JsonPerFeatureFormatterPlugin.kt
package com.nortal.test.core.plugin

import com.nortal.test.core.plugin.JsonPerFeatureReporter.readSources
import com.nortal.test.core.plugin.JsonPerFeatureReporter.addTestCase
import com.nortal.test.core.plugin.JsonPerFeatureReporter.startStep
import com.nortal.test.core.plugin.JsonPerFeatureReporter.finishStep
import com.nortal.test.core.plugin.JsonPerFeatureReporter.finishTestCase
import com.nortal.test.core.plugin.JsonPerFeatureReporter.addText
import com.nortal.test.core.plugin.JsonPerFeatureReporter.embed
import io.cucumber.plugin.EventListener
import io.cucumber.plugin.event.*

/**
 *
 * This class is a cucumber plugin that outputs tests results in a json format.
 * The main difference between this and standard json plugin is that this one will create a separate json file per feature.
 *
 *
 * Usage: plugin should be used as any other cucumber plugin by passing it to the runner.
 * <pre>`plugin = { "com.nortal.test.commons.plugin.JsonPerFeatureFormatterPlugin:buid/test-output/cucumber-report" }`</pre>
 *
 * Output folder should be provided by passing in a relative path as a parameter (the part after colon)
 */
class JsonPerFeatureFormatterPlugin(outputPath: String?) : EventListener {
    private val reporter: JsonPerFeatureReporter

    init {
        reporter = JsonPerFeatureReporter(outputPath)
    }

    override fun setEventPublisher(publisher: EventPublisher) {
        publisher.registerHandlerFor(TestSourceRead::class.java) { event: TestSourceRead? -> reporter.readSources(event) }
        publisher.registerHandlerFor(TestCaseStarted::class.java) { event: TestCaseStarted? ->
            reporter.addTestCase(
                event!!
            )
        }
        publisher.registerHandlerFor(TestStepStarted::class.java) { event: TestStepStarted? ->
            reporter.startStep(
                event!!
            )
        }
        publisher.registerHandlerFor(TestStepFinished::class.java) { event: TestStepFinished? ->
            reporter.finishStep(
                event!!
            )
        }
        publisher.registerHandlerFor(TestCaseFinished::class.java) { event: TestCaseFinished? ->
            reporter.finishTestCase(
                event!!
            )
        }
        publisher.registerHandlerFor(WriteEvent::class.java) { event: WriteEvent? ->
            reporter.addText(
                event!!
            )
        }
        publisher.registerHandlerFor(EmbedEvent::class.java) { event: EmbedEvent? ->
            reporter.embed(
                event!!
            )
        }
    }
}